File: kh1_src/tabs/journal_tabs/reports.py
from dash import Dash, html, dcc, callback, Input, Output, State, ALL
import kh1_src.kh1_utils as utils


def create_reports():
    kh1 = utils.kh1
    reports = [
        dcc.Checklist(
            options=[
                {"label": f"Ansem's Report {i*8+j+1}", "value": (1 << 7 - j)}\
                for j in range(8) if i*8+j+1 < 14
            ],
            value=[kh1.reports[i] & (1 << j) for j in range(8)],
            id={"type": "Report", "index": i},
        ) for i in range(len(kh1.reports))
    ]
    return html.Div(reports, style={"margin-top": 20})

@callback(
    Input({"type": "Report", "index": ALL}, "value"),
)
def reports_callback(values):
    kh1 = utils.kh1
    for i in range(len(kh1.reports)):
        kh1.reports[i] = sum(values[i])

File: kh1_src/tabs/journal_tabs/reports.js
export function create_reports() {
    const kh1jdiv = document.getElementById("kh1jdiv");
    let reports = `<div id="reports">`;
    for (let i = 0; i < 13; i++) {
        const idx = Math.floor(i / 8);
        const bit = i % 8;
        reports += `
        <label style="display: flex; align-items: center">
            <input
                type="checkbox"
                name=${idx}
                value=${1 << (7 - bit)}
                ${window.kh1.reports[idx] & (1 << (7 - bit)) ? "checked" : ""}
            >
            ${"Ansem's Report " + String(i + 1)}
        </label>`;
    }
    kh1jdiv.innerHTML = `
    <div>
        ${reports}
    </div>`;
    reports += "</div>";
    reports_callbacks();
}

function reports_callbacks() {
    const reports = document.getElementById("reports");
    reports.addEventListener("change", (e) => {
        if (e.target.checked)
            window.kh1.reports[e.target.name] |= e.target.value;
        else
            window.kh1.reports[e.target.name] &= ~e.target.value;
    });
}
